app/controllers/challenge_partnerships_controller.rb

```diff
@@ -24,14 +25,23 @@ def success; end
 
   def set_form
     token = params[:token] || params.dig(:challenge_partnership_form, :token)
-    notification_email = PartnershipNotificationEmail.find_by(token: token)
-    raise ActionController::RoutingError, "Not Found" if notification_email.blank?
+    partnership_id = params[:partnership] || params.dig(:challenge_partnership_form, :partnership)
+    if token.present?
+      notification_email = PartnershipNotificationEmail.find_by(token: token)
+      raise ActionController::RoutingError, "Not Found" if notification_email.blank?
+
+      @partnership = notification_email.partnership
+    elsif partnership_id.present?
+      @partnership = Partnership.find(partnership_id)
+      authorize(@partnership, :update?)
+    else
+      raise ActionController::RoutingError, "Not Found"
+    end
```

Comment:
This is getting pretty long now. Could we extract it into `@partnership = authorize_partnership` or even into a 
before_action? (then we don't need to use `and return` on redirections like on lines 42-43)